pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarPrimitiveSchemaHandler.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.sql.presto;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Date;
+
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.impl.schema.AutoConsumeSchema;
+import org.apache.pulsar.common.schema.SchemaInfo;
+
+/**
+ * A presto schema handler that interprets data using pulsar schema.
+ */
+public class PulsarPrimitiveSchemaHandler implements SchemaHandler {
+
+    private final SchemaInfo schemaInfo;
+    private final Schema<?> schema;
+
+    public PulsarPrimitiveSchemaHandler(SchemaInfo schemaInfo) {
+        this.schemaInfo = schemaInfo;
+        this.schema = AutoConsumeSchema.getSchema(schemaInfo);
+    }
+
+    @Override
+    public Object deserialize(ByteBuf byteBuf) {
+        byte[] data = ByteBufUtil.getBytes(byteBuf);
```

Comment:
I would recommend no allocating a new byte array every time we deserialize.  This could heavily degrade performance.  We should reuse pre-allocated buffers.  Please take a look at what is done in the JSONSchemaHandler:

https://github.com/apache/pulsar/blob/master/pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/JSONSchemaHandler.java#L58

Reply:
> I would recommend not allocating a new byte array every time we deserialize. This could heavily degrade performance. We should reuse pre-allocated buffers. Please take a look at what is done in the JSONSchemaHandler:
> 
> https://github.com/apache/pulsar/blob/master/pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/JSONSchemaHandler.java#L58

Pulsar primitive schema decode data by byte[] and can't use byte[] size. So we need to new a byte[], unless modify the method of primitive schema

Reply:
> Pulsar primitive schema decode data by byte[] and can't use byte[] size. So we need to new a byte[], unless modify the method of primitive schema

I don't quite follow.  Why can't you just do something like this

```
int size = payload.readableBytes();
        byte[] buffer = tmpBuffer.get();
        if (buffer.length < size) {
            // If the thread-local buffer is not big enough, replace it with
            // a bigger one
            buffer = new byte[size * 2];
            tmpBuffer.set(buffer);
        }

        payload.readBytes(buffer, 0, size);

       schema.decode(data);
```

Reply:
> > Pulsar primitive schema decode data by byte[] and can't use byte[] size. So we need to new a byte[], unless modify the method of primitive schema
> 
> I don't quite follow. Why can't you just do something like this
> 
> ```
> int size = payload.readableBytes();
>         byte[] buffer = tmpBuffer.get();
>         if (buffer.length < size) {
>             // If the thread-local buffer is not big enough, replace it with
>             // a bigger one
>             buffer = new byte[size * 2];
>             tmpBuffer.set(buffer);
>         }
> 
>         payload.readBytes(buffer, 0, size);
> 
>        schema.decode(data);
> ```

For example,  StringSchema decode method :
```
    public String decode(byte[] bytes) {
        if (null == bytes) {
            return null;
        } else {
            return new String(bytes, charset);
        }
    }
```
so we need to overload the decode method to allow the size of the readable byte [] to be passed in.

Reply:
I see.  I think you can just use byteBuf.array().  I looked at the code and it seems to just return the underlying byte array.

Reply:
> I see. I think you can just use byteBuf.array(). I looked at the code and it seems to just return the underlying byte array.

but not all ByteBuf implementation classes implement this method.

Reply:
in general we should just improve our schema interface to take `ByteBuf` as the bytes array. we shouldn't do this profiling separately in presto plugin. this is against the purpose of moving presto to use `Schema` interface as I started this change. 

I would suggest creating a follow up issue to add methods in Schema to support deserializing from `ByteBuf` and we address the problem in Schema implementations, and then change the code here.

Reply:
@sijie can you create an follow on issue then?

Reply:
https://github.com/apache/pulsar/issues/4775